routes/runtime/github.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+class CloverRuntime
+  hash_branch("github") do |r|
+    if (runner = GithubRunner[vm_id: @vm.id]).nil? || (repository = runner.repository).nil?
+      fail CloverError.new(401, "InvalidCredentials", "There was an error logging in")
+    end
+
+    r.get "cache" do
+      fail CloverError.new(204, "NotFound", "No cache entry") if repository.access_key.nil?
+
+      keys, version = r.params["keys"]&.split(","), r.params["version"]
+      fail CloverError.new(400, "InvalidRequest", "Wrong parameters") if keys.nil? || keys.empty? || version.nil?
+
+      scopes = [runner.workflow_job&.dig("head_branch"), repository.default_branch].compact
+      entry = repository.cache_entries_dataset
+        .exclude(committed_at: nil)
+        .where(key: keys, version: version, scope: scopes).all
+        .min_by { keys.index(_1.key) + (scopes.index(_1.scope) * keys.size) }
+
+      fail CloverError.new(204, "NotFound", "No cache entry") if entry.nil?
+
+      entry.update(last_accessed_at: Time.now, last_accessed_by: runner.id)
+      signed_url, _ = repository.url_presigner.presigned_request(:get_object, bucket: repository.bucket_name, key: entry.blob_key)
```

Comment:
For how long this key is valid? I think we should set the validation duration to a minimum reasonable amount to prevent possibility of using our blob storage as file share by malicious actor.

Reply:
Yes, Users can extract it since we return it to the runner VM. The default time is set to 900 seconds (15 minutes), which I believe provides sufficient time for cache upload/download. This is based on our measurements that uploading or downloading a 10GB cache takes no more than 7 minutes with our current blob storage implementation. 

I also believe that a 15-minute limit is a good security measure. What are your thoughts on this? I've explicitly added this argument, even though it matches the default value. 

Given that the presigned URL is only valid for a specific bucket and prefix, the potential for attack is relatively small. However, there is always the risk of an attacker attempting to poison the caches. For this reason, GitHub does not permit branches to access caches from other branches.